### PR TITLE
avoid double Object.defineProperty for native descriptors

### PR DESCRIPTION
**Pull request by bekzod**

### CODE REVIEW

File: packages/ember-metal/lib/properties.ts [hunk 1]
```diff
@@ -188,16 +217,18 @@ export function defineProperty(
...
 
-    Object.defineProperty(obj, keyName, {
-      configurable: true,
-      enumerable,
-      get: DESCRIPTOR_GETTER_FUNCTION(keyName, value),
-    });
-
-    meta.writeDescriptors(keyName, value);
-
-    if (typeof desc.setup === 'function') {
```
Comment (bekzod) [inline packages/ember-metal/lib/properties.ts]: in setup `NativeDescriptor` would make another `Object.defineProperty` call

File: packages/ember-metal/lib/properties.ts [hunk 2]
```diff
@@ -188,16 +217,18 @@ export function defineProperty(
...
-      desc.setup(obj, keyName);
+    if (desc instanceof NativeDescriptor) {
+      Object.defineProperty(obj, keyName, desc.desc);
+    } else {
+      Object.defineProperty(obj, keyName, {
+        configurable: true,
+        enumerable,
+        get: DESCRIPTOR_GETTER_FUNCTION(keyName, value),
+      });
+      meta.writeDescriptors(keyName, value);
```
Comment (rwjblue) [inline packages/ember-metal/lib/properties.ts]: The `meta.writeDescriptors` here was previously done in both cases and now its only done for non-native.

Comment (bekzod) [reply packages/ember-metal/lib/properties.ts]: Yes sorry, was going to ask if that is bad idea, else will investigate, tests don’t fail though

Comment (rwjblue) [reply packages/ember-metal/lib/properties.ts]: To be honest, I am unsure if it is important, but I would assume that we don't want to change the results of things like `descriptorFor` and whatnot.

**Comment by rwjblue** [comment]
I just created a merge conflict here (by merging #16739 ), can you rebase / fix up the conflicts?

**Review by rwjblue — CHANGES_REQUESTED** [review]
After chatting with @krisselden about this, I think we'd rather fix the double `Object.defineProperty` by moving _all_ of the descriptor define properties to be done in `Descriptor.prototype.setup`..

**Comment by bekzod** [comment]
done

File: packages/ember-metal/lib/properties.ts [hunk 3]
```diff
@@ -46,6 +50,36 @@ export abstract class Descriptor {
...
 
+/**
+  A wrapper for a native ES5 descriptor. In an ideal world, we wouldn't need
+  this at all, however, the way we currently flatten/merge our mixins require
+  a special value to denote a descriptor.
+
+  @class NativeDescriptor
+  @private
+*/
+export class NativeDescriptor extends Descriptor {
```
Comment (rwjblue) [inline packages/ember-metal/lib/properties.ts]: Given how this works now, I think we can move this back to its original location (in `descriptor.ts`), right?

**Review by rwjblue — CHANGES_REQUESTED** [review]
Update looks nice, only one small change to make and this seems good to me...

File: packages/ember-metal/lib/properties.ts (hunk 3 above)
Comment (bekzod) [reply packages/ember-metal/lib/properties.ts]: yep, just thought about avoiding extra import/exports :P

Comment (bekzod) [reply packages/ember-metal/lib/properties.ts]: done

**Comment by rwjblue** [comment]
restarted ci

**Review by rwjblue — APPROVED** [review]

**Comment by krisselden** [comment]
this is failing in Edge

image

**Comment by krisselden** [comment]
@bekzod this looks good, are you investigating the CI failure?

**Comment by bekzod** [comment]
briefly looked have no idea :P, will investigate more later when I have time :)

**Comment by rwjblue** [comment]
restarted the browserstack job

**Comment by krisselden** [comment]
I’d love to merge this, anyone able to investigate the IE failure?

**Comment by bekzod** [comment]
there is quirk with Edge where `non-configurable property` do not throw when `delete`ed
Edge 17: https://cl.ly/9f5bfc86f5d7
IE11: https://cl.ly/a62f4c3a0fa7
fixed test by adding flag for Edge case 😝

**Comment by bekzod** [comment]
needs restart, other than that it should work

**Comment by bekzod** [comment]
ping @krisselden

**Review by krisselden — APPROVED** [review]
